rac2: implement stream token counter provider

kvoli opened the pull request:
First two commits from https://github.com/cockroachdb/cockroach/pull/128217

---

Replace the existing `StreamTokenCounterProvider` interface with a
struct implementation. `StreamTokenCounterProvider` provides two
methods, `Send` and `Eval`, which should be used to access a (eval|send)
token counter for a given stream on a node.

Resolves: https://github.com/cockroachdb/cockroach/issues/128027
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128221)
<!-- Reviewable:end -->

kvoli committed:
rac2: implement stream token counter provider

Replace the existing `StreamTokenCounterProvider` interface with a
struct implementation. `StreamTokenCounterProvider` provides two
methods, `Send` and `Eval`, which should be used to access a (eval|send)
token counter for a given stream on a node.

Resolves: #128027
Release note: None

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128221#-:-O3ecWNS-PwBSByzzzqp:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist and @pav-kv)



<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go:
```diff
@@ -14,17 +14,55 @@ import (
+}
+
+// Eval returns the evaluation token counter for the given stream.
+func (stcp *StreamTokenCounterProvider) Eval(stream kvflowcontrol.Stream) TokenCounter {
```
nit: one-letter receiver name is generally ok, like `s` or `p` in this case.

kvoli reviewed (commented):
TYFTRS!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128221#-:-O3jDl3K-RsceCbTog6i:bx5okfc)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128221#gh-1706950271:-O3jCOwPD7mRP5ozGOGj:brzbbhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8658435b43252b24ef8ca54541af9567713ac7/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L26)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Is clock unused?
</blockquote></details>

It is, since we removed the clock in token counter. Removed here as well.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128221#gh-1706959164:-O3jCV4e3MQeDA5ut106:b-56297) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8658435b43252b24ef8ca54541af9567713ac7/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L30)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Nobody seems to be deleting from these maps. Add a TODO to fix that?
</blockquote></details>

Good idea, added a TODO above the struct.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128221#gh-1706955967:-O3jCg0F6Y6q6jGfFlSe:b-sen88f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8658435b43252b24ef8ca54541af9567713ac7/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L31)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Use `syncutil.Map` or TODO to check if provides better performance?
</blockquote></details>

We probably won't expect too frequent access, given this is used on range controller init. I added a TODO anyway.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128221#gh-1706963643:-O3jCyuC7SGlyVuuGIr5:b-7qqb55) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8658435b43252b24ef8ca54541af9567713ac7/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L45)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

nit: one-letter receiver name is generally ok, like `s` or `p` in this case.
</blockquote></details>

I'm going to leave as is, the send token watcher below is meatier in comparison -- so I'm leaving the `s` open for it.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128221#gh-1706948465:-O3jD6TB0CmSmWuzGKoc:b-sk7t7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8658435b43252b24ef8ca54541af9567713ac7/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L53)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

The happy case accessed the map twice. Consider inverting it: if found, return; otherwise, create/insert and return.
Same bellow.
</blockquote></details>

Nice idea, updated.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (approved):
LGTM % remaining nits and rebasing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128221#-:-O3lX9Ay2qh9m2aW2D3L:be4ojai)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist, @kvoli, and @sumeerbhola)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128221#gh-1706963643:-O3lVmFT8_apDT1L-vG7:b-fvkja4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8658435b43252b24ef8ca54541af9567713ac7/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L45)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

I'm going to leave as is, the send token watcher below is meatier in comparison -- so I'm leaving the `s` open for it.
</blockquote></details>

Why worry about cross-type naming of the receivers? This is a local var. Also, `p` for "provider" works too if there is something in your workflow requiring uniqueness of names.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 25 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/128221#-O3lWU2s9GTOcVLEkYCi:-O3lWU2s9GTOcVLEkYCj:bgeffx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c112ea6f8ad3bec3ac5d30305a695f455a113bdf/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L25)):*
> ```Go
> //
> // TODO(kvoli): Add stream deletion upon decommissioning a store.
> // TODO(kvoli): Check mutex performance against sync.Map.
> ```

nit: syncutil.Map

#127079
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 57 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/128221#-O3lV_pc-V20ZC1KwOBF:-O3lV_pc-V20ZC1KwOBG:b-2b2sk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c112ea6f8ad3bec3ac5d30305a695f455a113bdf/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L57)):*
> ```Go
> 
> // Send returns the send token counter for the given stream.
> func (sstc *StreamTokenCounterProvider) Send(stream kvflowcontrol.Stream) TokenCounter {
> ```

The receiver name mismatches here and above. Make them consistent? Might as well make it short :)


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
TYFTR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128221#-:-O3o5o9rANjRhpZtCHNn:b-80sls6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist, @pav-kv, and @sumeerbhola)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/128221#gh-1706963643:-O3o5qNvELsNWYSmJYDr:b5k6ioi) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8658435b43252b24ef8ca54541af9567713ac7/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L45)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Why worry about cross-type naming of the receivers? This is a local var. Also, `p` for "provider" works too if there is something in your workflow requiring uniqueness of names.
</blockquote></details>

Changed to p.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 25 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/128221#-O3lWU2s9GTOcVLEkYCi:-O3o5p3ECVt76KNiMo49:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c112ea6f8ad3bec3ac5d30305a695f455a113bdf/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L25)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

nit: syncutil.Map

#127079
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 57 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/128221#-O3lV_pc-V20ZC1KwOBF:-O3o5pg35GR_y1W_0bIx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c112ea6f8ad3bec3ac5d30305a695f455a113bdf/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L57)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

The receiver name mismatches here and above. Make them consistent? Might as well make it short :)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kvoli commented:
TYFTRs!

bors r=pav-kv,sumeerbhola

kvoli commented:
rebased on master.

kvoli commented:
bors r=pav-kv,sumeerbhola
